# Kaggle Competition House Prices
By: Rohinesh Ram

# 0. Imports / Functions

In [15]:
import pandas as pd
import numpy as np

# 1. Load the Data

- two data sets from Kaggle, testset and trainset
- also txt of feature descriptions

In [53]:
# Train data set
df_train = pd.read_csv('data_folder/train.csv').fillna('NA') # replace NAN with NA
# Test data set, replace NaN with NA because pandas reas NA as NaN
df_test = pd.read_csv('data_folder/test.csv').fillna('NA') 
# Data description
f = open("data_folder/data_description.txt", "r")

# Display each
print(f.read())
df_train.head()
df_test.head()

MSSubClass: Identifies the type of dwelling involved in the sale.	

        20	1-STORY 1946 & NEWER ALL STYLES
        30	1-STORY 1945 & OLDER
        40	1-STORY W/FINISHED ATTIC ALL AGES
        45	1-1/2 STORY - UNFINISHED ALL AGES
        50	1-1/2 STORY FINISHED ALL AGES
        60	2-STORY 1946 & NEWER
        70	2-STORY 1945 & OLDER
        75	2-1/2 STORY ALL AGES
        80	SPLIT OR MULTI-LEVEL
        85	SPLIT FOYER
        90	DUPLEX - ALL STYLES AND AGES
       120	1-STORY PUD (Planned Unit Development) - 1946 & NEWER
       150	1-1/2 STORY PUD - ALL AGES
       160	2-STORY PUD - 1946 & NEWER
       180	PUD - MULTILEVEL - INCL SPLIT LEV/FOYER
       190	2 FAMILY CONVERSION - ALL STYLES AND AGES

MSZoning: Identifies the general zoning classification of the sale.
		
       A	Agriculture
       C	Commercial
       FV	Floating Village Residential
       I	Industrial
       RH	Residential High Density
       RL	Residential Low Density
       RP	Residential Low Density Park 
       RM

Id  MSSubClass MSZoning LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH        80.0    11622   Pave    NA      Reg   
1  1462          20       RL        81.0    14267   Pave    NA      IR1   
2  1463          60       RL        74.0    13830   Pave    NA      IR1   
3  1464          60       RL        78.0     9978   Pave    NA      IR1   
4  1465         120       RL        43.0     5005   Pave    NA      IR1   

  LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence MiscFeature  \
0         Lvl    AllPub  ...         120        0     NA  MnPrv          NA   
1         Lvl    AllPub  ...           0        0     NA     NA        Gar2   
2         Lvl    AllPub  ...           0        0     NA  MnPrv          NA   
3         Lvl    AllPub  ...           0        0     NA     NA          NA   
4         HLS    AllPub  ...         144        0     NA     NA          NA   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  
0       0      6    2010        WD         Normal  
1   12500      6    2010        WD         Normal  
2       0      3    2010        WD         Normal  
3       0      6    2010        WD         Normal  
4       0      1    2010        WD         Normal  

[5 rows x 80 columns]

## 1.1 Prepare Feature matrix and Target Vector

Our target vector is the 'SalePrice' column.

In [50]:
y_train = df_train['SalePrice']
X_train = df_train.loc[:,'MSSubClass':'SaleCondition'] # do not need ID
X_test = df_test.loc[:,'MSSubClass':'SaleCondition'] # do not need ID
id = df_train['Id']

print("X train shape:", X_train.shape)
print("y train shape:", y_train.shape)
print("X test shape:", X_test.shape)
X_test.columns

X train shape: (1460, 79)
y train shape: (1460,)
X test shape: (1459, 79)


Index(['MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street', 'Alley',
       'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope',
       'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle',
       'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'RoofStyle',
       'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea',
       'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond',
       'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2',
       'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating', 'HeatingQC',
       'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF',
       'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath',
       'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual', 'TotRmsAbvGrd',
       'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType', 'GarageYrBlt',
       'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond',
       'PavedDrive', 'Wo

The feature matrix contains 80 features. The train and test sets contain 1460 and 1459 data points respectively

## 2. Inspect and Preprocess

We have a mix of numerical, categorical features, and ordinal features, lets find which are which:

**Numerical**: LotFrontage, LotArea, YearBuilt, YearRemodAdd, MasVnrArea, BsmtFinSF1, BsmtFinSF2, BsmtUnfSF, TotalBsmtSF, 1stFlrSF, 2ndFlrSF, LowQualFinSF, GrLivArea, BsmtFullBath, BsmtHalfBath, FullBath, HalfBath, Bedroom, Kitchen, TotRmsAbvGrd, Fireplaces, GarageYrBlt, GarageCars, GarageArea, WoodDeckSF, OpenPorchSF, EnclosedPorch, 3SsnPorch, ScreenPorch, PoolArea, MiscVal, MoSold, YrSold 

**Categorical**: MSSubClass, MSZoning, Street, Alley, LotShape, LandContour, Utilities, LotConfig, LandSlope, Neighborhood, Condition1, Condition2, BldgType, HouseStyle, RoofStyle, RoofMatl, Exterior1st, Exterior2nd, MasVnrType, Foundation, Heating, CentralAir, Electrical, Functional, GarageType, GarageFinish, PavedDrive, MiscFeature, SaleType, SaleCondition

Features related to quality or condition were considered ordinal because they imply some sort of rank or order, i.e. good quality is better than fair quality.

**Ordinal**: OverallQual, OverallCond, ExterQual, ExterCond, BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1, BsmtFinType2, HeatingQC, KitchenQual, FireplaceQu, GarageQual, GarageCond, PoolQC, Fence

In [51]:
numerical = 'LotFrontage, LotArea, YearBuilt, YearRemodAdd, MasVnrArea, BsmtFinSF1, BsmtFinSF2, BsmtUnfSF, TotalBsmtSF, 1stFlrSF, 2ndFlrSF, LowQualFinSF, GrLivArea, BsmtFullBath, BsmtHalfBath, FullBath, HalfBath, BedroomAbvGr, KitchenAbvGr, TotRmsAbvGrd, Fireplaces, GarageYrBlt, GarageCars, GarageArea, WoodDeckSF, OpenPorchSF, EnclosedPorch, 3SsnPorch, ScreenPorch, PoolArea, MiscVal, MoSold, YrSold'
categorical = 'MSSubClass, MSZoning, Street, Alley, LotShape, LandContour, Utilities, LotConfig, LandSlope, Neighborhood, Condition1, Condition2, BldgType, HouseStyle, RoofStyle, RoofMatl, Exterior1st, Exterior2nd, MasVnrType, Foundation, Heating, CentralAir, Electrical, Functional, GarageType, GarageFinish, PavedDrive, MiscFeature, SaleType, SaleCondition'
ordinal = 'OverallQual, OverallCond, ExterQual, ExterCond, BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1, BsmtFinType2, HeatingQC, KitchenQual, FireplaceQu, GarageQual, GarageCond, PoolQC, Fence'

numerical = numerical.split(', ')
categorical = categorical.split(', ')
ordinal = ordinal.split(', ')

## Pre-process

Convert categorical, and ordinal to numerical using OneHotEncoder and OrdinalEncoder

In [59]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OrdinalEncoder

ct = ColumnTransformer(
    [('onehot', OneHotEncoder(sparse = False), categorical),
    ('ordinal', OrdinalEncoder(), ordinal),
    ('nothing', 'passthrough', numerical)])

ct.fit(X_train)
X_train_trans = ct.transform(X_train)
print("Shape of Feature Matrix after encoding:",X_train_trans.shape)

Shape of Feature Matrix after encoding: (1460, 258)
